ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -66,48 +156,89 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
-      }
-      uuid = Utils.readIntString(stream);
-      if (ensureFullyRead && stream.read() != -1) {
-        throw new IllegalArgumentException("Stream should have no more available bytes to read");
-      }
-    } else {
-      throw new IllegalArgumentException("version " + version + " not supported for blob id");
-    }
-  }
-
   public short sizeInBytes() {
-    return (short) (Version_Size_In_Bytes + partitionId.getBytes().length + UUID_Size_In_Bytes + uuid.length());
+    switch (version) {
+      case BLOB_ID_V1:
+        return (short) (VERSION_SIZE_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_IN_BYTES + uuid.length());
+      case BLOB_ID_V2:
+        return (short) (VERSION_SIZE_IN_BYTES + FLAG_SIZE_IN_BYTES + DATACENTER_ID_SIZE_IN_BYTES
+            + ACCOUNT_ID_SIZE_IN_BYTES + CONTAINER_ID_SIZE_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_IN_BYTES
+            + uuid.length());
+      default:
+        throw new IllegalArgumentException("blob id version=" + version + " not supported");
+    }
   }
 
+  /**
+   * Gets the {@link PartitionId} this blob belongs to.
+   * @return The {@link PartitionId}.
+   */
   public PartitionId getPartition() {
     return partitionId;
   }
 
+  /**
+   * Gets the id of the {@link Account} who created this blob. If this information was not available when the
+   * blob id was formed, it will return -1.
+   * @return The id of the {@link Account} who created this blob.
+   */
+  public short getAccountId() {
+    return accountId;
+  }
+
+  /**
+   * Gets the id of the {@link Container} where this blob belongs to. If this information was not available when
+   * the blob id was formed, it will return -1.
+   * @return The id of the {@link Container} where this blob belongs to.
+   */
+  public short getContainerId() {
+    return containerId;
+  }
+
+  /**
+   * Gets the id of the datacenter where this blob was originally posted. If this information was not available
+   * when the blob id was formed, it will return -1.
+   * @return The id of the datacenter where this blob was originally posted.
+   */
+  public short getDatacenterId() {
+    return datacenterId;
+  }
+
+  /**
+   * Gets the flag metadata of this blob id. If this information was not available when the blob id was formed, it
+   * will return 0.
+   * @return The flag of the blobId.
+   */
+  public short getFlag() {
+    return flag;
+  }
+
   @Override
   public byte[] toBytes() {
     ByteBuffer idBuf = ByteBuffer.allocate(sizeInBytes());
     idBuf.putShort(version);
+    switch (version) {
+      case BLOB_ID_V1:
+        break;
+
+      case BLOB_ID_V2:
+        idBuf.put(flag);
+        idBuf.put(datacenterId);
+        idBuf.putShort(accountId);
+        idBuf.putShort(containerId);
+        break;
+
+      default:
+        throw new IllegalArgumentException("blob id version=" + version + " not supported");
+    }
     idBuf.put(partitionId.getBytes());
     idBuf.putInt(uuid.getBytes().length);
```

Comment:
Have changed toe doc to short. After discussion with Priyesh, it was decided to keep using int not to introduce confusion on the same field used in two versions.